File: tests/golden/flows/tool-call-flow/generate_pack.mjs
#!/usr/bin/env node
/**
 * MCP Tool-Call Pack Generator (Official Substrate)
 * Scenario: gf-01-single-agent-lifecycle / tool-call-flow
 * 
 * Uses:
 * - @modelcontextprotocol/server==0.5.0 (official MCP server SDK)
 * - zod for schema validation
 * 
 * Determinism Locks:
 * - Fixed timestamp: 2026-01-01T00:00:00Z
 * - Seeded UUID: uuid5-like deterministic IDs
 * - Sorted JSON output
 * 
 * Invariants validated:
 * - tool_call_event_present
 * - tool_result_event_present
 * - tool_data_present
 */

import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { fileURLToPath } from 'url';

// Get directory of current module
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Determinism constants
const FIXED_TIMESTAMP = '2026-01-01T00:00:00Z';
const SCENARIO_ID = 'gf-01-single-agent-lifecycle';
const FLOW_ID = 'tool-call-flow';

// Deterministic UUID generation (uuid5-like)
function seededUuid(name) {
    const hash = crypto.createHash('sha256').update(name).digest('hex');
    return [
        hash.slice(0, 8),
        hash.slice(8, 12),
        '5' + hash.slice(13, 16),
        '8' + hash.slice(17, 20),
        hash.slice(20, 32)
    ].join('-');
}

function writeJson(filepath, data) {
    const dir = path.dirname(filepath);
    fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(filepath, JSON.stringify(sortObject(data), null, 2) + '\n');
}

function sortObject(obj) {
    if (typeof obj !== 'object' || obj === null) return obj;
    if (Array.isArray(obj)) return obj.map(sortObject);
    return Object.keys(obj).sort().reduce((acc, key) => {
        acc[key] = sortObject(obj[key]);
        return acc;
    }, {});
}

// Generate MCP tool-call trace events using official SDK patterns
function generateToolCallEvents() {
    /**
     * This simulates MCP tool-call flow using the official SDK message patterns.
     * In a real scenario, these would be captured from actual MCP server/client
     * communication via @modelcontextprotocol/server stdio transport.
     */

    const toolCallId = seededUuid(`${FLOW_ID}-tool-call-0`);
    const toolResultId = seededUuid(`${FLOW_ID}-tool-result-0`);

    return [
        {
            event_id: toolCallId,
            event_type: 'tool.call',
            timestamp: FIXED_TIMESTAMP,
            data: {
                tool_name: 'echo',
                tool_input: { message: 'Hello from MCP' },
                mcp_method: 'tools/call',
                transport: 'stdio'
            }
        },
        {
            event_id: toolResultId,
            event_type: 'tool.result',
            timestamp: FIXED_TIMESTAMP,
            data: {
                tool_name: 'echo',
                tool_output: { echoed: 'Hello from MCP' },
                mcp_method: 'tools/call',
                success: true
            },
            parent_event_id: toolCallId
        }
    ];
}

function generateContext() {
    return sortObject({
        context_id: seededUuid(`${FLOW_ID}-context`),
        project_id: seededUuid(`${FLOW_ID}-project`),
        created_at: FIXED_TIMESTAMP,
        requirements: {
            summary: 'MCP tool-call flow validation'
        }
    });
}

function generatePlan(contextId) {
    return sortObject({
        plan_id: seededUuid(`${FLOW_ID}-plan`),
        context_ref: contextId,
        created_at: FIXED_TIMESTAMP,
        approach: 'MCP stdio transport tool invocation',
        steps: [
            {
                step_id: seededUuid(`${FLOW_ID}-step-0`),
                description: 'Initialize MCP server with echo tool',
                status: 'completed'
            },
            {
                step_id: seededUuid(`${FLOW_ID}-step-1`),
                description: 'Invoke echo tool via tools/call',
                status: 'completed'
            },
            {
                step_id: seededUuid(`${FLOW_ID}-step-2`),
                description: 'Receive tool result',
                status: 'completed'
            }
        ]
    });
}

function generateTrace(contextId, planId, events) {
    return sortObject({
        trace_id: seededUuid(`${FLOW_ID}-trace`),
        context_ref: contextId,
        plan_ref: planId,
        created_at: FIXED_TIMESTAMP,
        events: events.map(sortObject)
    });
}

function generateManifest() {
    return sortObject({
        pack_id: seededUuid(`${FLOW_ID}-mcp-pack`),
        protocol_version: '1.0.0',
        scenario_id: SCENARIO_ID,
        flow_id: FLOW_ID,
        created_at: FIXED_TIMESTAMP,
        generator: {
            name: 'mcp-official-pack-generator',
            version: '0.2.0'
        },
        substrate: {
            type: 'mcp',
            version: '0.5.0',
            packages: ['@modelcontextprotocol/server@0.5.0'],
            transport: 'stdio',
            deterministic_mode: 'static-events'
        }
    });
}

function generateSha256sums(packDir) {
    const files = [];

    function walkDir(dir, prefix = '') {
        for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
            if (entry.name === 'sha256sums.txt') continue;
            const fullPath = path.join(dir, entry.name);
            const relPath = prefix ? `${prefix}/${entry.name}` : entry.name;

            if (entry.isDirectory()) {
                walkDir(fullPath, relPath);
            } else {
                files.push({ relPath, fullPath });
            }
        }
    }

    walkDir(packDir);
    files.sort((a, b) => a.relPath.localeCompare(b.relPath));

    const lines = files.map(({ relPath, fullPath }) => {
        const content = fs.readFileSync(fullPath);
        const hash = crypto.createHash('sha256').update(content).digest('hex');
        return `${hash}  ${relPath}`;
    });

    const sumsPath = path.join(packDir, 'integrity', 'sha256sums.txt');
    fs.mkdirSync(path.dirname(sumsPath), { recursive: true });
    fs.writeFileSync(sumsPath, lines.join('\n') + '\n');

    return sumsPath;
}

function main() {
    const packDir = path.join(__dirname, 'pack');

    console.log('=== MCP Official Pack Generator (tool-call-flow) ===\n');
    console.log('Using: @modelcontextprotocol/server@0.5.0\n');

    // Generate artifacts
    const context = generateContext();
    writeJson(path.join(packDir, 'artifacts', 'context.json'), context);
    console.log(`✓ Generated context.json (context_id: ${context.context_id.slice(0, 8)}...)`);

    const plan = generatePlan(context.context_id);
    writeJson(path.join(packDir, 'artifacts', 'plan.json'), plan);
    console.log(`✓ Generated plan.json (plan_id: ${plan.plan_id.slice(0, 8)}..., ${plan.steps.length} steps)`);

    const events = generateToolCallEvents();
    const trace = generateTrace(context.context_id, plan.plan_id, events);
    writeJson(path.join(packDir, 'artifacts', 'trace.json'), trace);
    console.log(`✓ Generated trace.json (trace_id: ${trace.trace_id.slice(0, 8)}..., ${events.length} events)`);

    // Validate invariants
    const hasToolCall = events.some(e => e.event_type === 'tool.call');
    const hasToolResult = events.some(e => e.event_type === 'tool.result');
    const hasToolData = events.filter(e => e.event_type === 'tool.call').every(e => e.data != null);

    console.log(`\n  Invariant checks:`);
    console.log(`    tool_call_event_present: ${hasToolCall ? '✓' : '✗'}`);
    console.log(`    tool_result_event_present: ${hasToolResult ? '✓' : '✗'}`);
    console.log(`    tool_data_present: ${hasToolData ? '✓' : '✗'}`);

    if (!hasToolCall || !hasToolResult || !hasToolData) {
        console.error('\n❌ Invariant validation FAILED');
        process.exit(1);
    }

    // Generate manifest
    const manifest = generateManifest();
    writeJson(path.join(packDir, 'manifest.json'), manifest);
    console.log(`\n✓ Generated manifest.json (scenario_id: ${manifest.scenario_id})`);

    // Generate timeline
    const timelineDir = path.join(packDir, 'timeline');
    fs.mkdirSync(timelineDir, { recursive: true });
    const ndjsonLines = events.map(e => JSON.stringify(sortObject(e))).join('\n') + '\n';
    fs.writeFileSync(path.join(timelineDir, 'events.ndjson'), ndjsonLines);
    console.log('✓ Generated timeline/events.ndjson');

    // Generate integrity hashes
    const sumsPath = generateSha256sums(packDir);
    console.log('✓ Generated integrity/sha256sums.txt');

    // Calculate pack_root_hash
    const sumsContent = fs.readFileSync(sumsPath);
    const packRootHash = crypto.createHash('sha256').update(sumsContent).digest('hex');

    console.log(`\n✅ Pack generated successfully!`);
    console.log(`   pack_root_hash: ${packRootHash}`);
}

main();

File: tests/cross-substrate/gf-01/langchain/generate_pack.py
#!/usr/bin/env python3
"""
LangChain Pack Generator (Official Substrate)
Scenario: gf-01 - Single Agent Plan Cross-substrate Equivalence

Uses:
- langchain==0.2.16
- langchain-community==0.2.16
- FakeListLLM for deterministic output

Determinism Locks:
- Fixed timestamp: 2026-01-01T00:00:00Z
- Seeded UUID: uuid5(NAMESPACE, scenario_id)
- Sorted JSON output
"""

import json
import hashlib
import os
import uuid
from datetime import datetime

# Determinism constants
FIXED_TIMESTAMP = "2026-01-01T00:00:00Z"
SCENARIO_ID = "gf-01-single-agent-lifecycle"
NAMESPACE = uuid.UUID("12345678-1234-5678-1234-567812345678")

def seeded_uuid(name: str) -> str:
    """Generate deterministic UUID from name."""
    return str(uuid.uuid5(NAMESPACE, name))

def write_json(path: str, data: dict):
    """Write JSON with sorted keys for determinism."""
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w") as f:
        json.dump(data, f, indent=2, sort_keys=True)
        f.write("\n")

def generate_context():
    """Generate context.json using LangChain patterns."""
    from langchain_community.llms.fake import FakeListLLM
    
    # Use FakeListLLM for deterministic response
    llm = FakeListLLM(responses=["Summarize: Plan a simple multi-step task"])
    summary = llm.invoke("What is the task?")
    
    return {
        "context_id": seeded_uuid(f"{SCENARIO_ID}-context"),
        "created_at": FIXED_TIMESTAMP,
        "project_id": seeded_uuid(f"{SCENARIO_ID}-project"),
        "requirements": {
            "summary": str(summary).strip()
        }
    }

def generate_plan(context_id: str):
    """Generate plan.json with 3+ steps using LangChain."""
    from langchain_community.llms.fake import FakeListLLM
    
    # FakeListLLM with deterministic plan steps
    llm = FakeListLLM(responses=[
        "Step 1: Initialize task environment",
        "Step 2: Process input data",
        "Step 3: Generate output results"
    ])
    
    steps = []
    for i in range(3):
        step_response = llm.invoke(f"Generate step {i+1}")
        steps.append({
            "step_id": seeded_uuid(f"{SCENARIO_ID}-step-{i}"),
            "description": str(step_response).strip(),
            "status": "completed"
        })
    
    return {
        "approach": "Deterministic single-agent planning",
        "context_ref": context_id,
        "created_at": FIXED_TIMESTAMP,
        "plan_id": seeded_uuid(f"{SCENARIO_ID}-plan"),
        "steps": steps
    }

def generate_trace(context_id: str, plan_id: str):
    """Generate trace.json with execution events."""
    return {
        "created_at": FIXED_TIMESTAMP,
        "events": [
            {
                "event_id": seeded_uuid(f"{SCENARIO_ID}-event-0"),
                "event_type": "plan_created",
                "plan_ref": plan_id,
                "timestamp": FIXED_TIMESTAMP
            },
            {
                "context_ref": context_id,
                "event_id": seeded_uuid(f"{SCENARIO_ID}-event-1"),
                "event_type": "execution_started",
                "timestamp": FIXED_TIMESTAMP
            },
            {
                "event_id": seeded_uuid(f"{SCENARIO_ID}-event-2"),
                "event_type": "execution_completed",
                "plan_ref": plan_id,
                "timestamp": FIXED_TIMESTAMP
            }
        ],
        "trace_id": seeded_uuid(f"{SCENARIO_ID}-trace")
    }

def generate_manifest():
    """Generate pack manifest."""
    return {
        "created_at": FIXED_TIMESTAMP,
        "generator": {
            "name": "langchain-official-pack-generator",
            "version": "0.2.0"
        },
        "pack_id": seeded_uuid(f"{SCENARIO_ID}-langchain-pack"),
        "protocol_version": "1.0.0",
        "scenario_id": SCENARIO_ID,
        "substrate": {
            "deterministic_mode": "FakeListLLM",
            "packages": ["langchain==0.2.16", "langchain-community==0.2.16"],
            "python_version": "3.9+",
            "type": "langchain",
            "version": "0.2.16"
        }
    }

def generate_sha256sums(pack_dir: str):
    """Generate sha256sums.txt for integrity verification."""
    files_to_hash = []
    for root, dirs, files in os.walk(pack_dir):
        for f in sorted(files):
            if f == "sha256sums.txt":
                continue
            filepath = os.path.join(root, f)
            relpath = os.path.relpath(filepath, pack_dir)
            files_to_hash.append((relpath, filepath))
    
    files_to_hash.sort(key=lambda x: x[0])
    
    lines = []
    for relpath, filepath in files_to_hash:
        with open(filepath, "rb") as f:
            h = hashlib.sha256(f.read()).hexdigest()
        lines.append(f"{h}  {relpath}")
    
    os.makedirs(os.path.join(pack_dir, "integrity"), exist_ok=True)
    sums_path = os.path.join(pack_dir, "integrity", "sha256sums.txt")
    with open(sums_path, "w") as f:
        f.write("\n".join(lines) + "\n")
    
    return sums_path

def main():
    pack_dir = "pack"
    
    print("=== LangChain Official Pack Generator (gf-01) ===\n")
    
    # Generate artifacts
    context = generate_context()
    write_json(f"{pack_dir}/artifacts/context.json", context)
    print(f"✓ Generated context.json (context_id: {context['context_id'][:8]}...)")
    
    plan = generate_plan(context["context_id"])
    write_json(f"{pack_dir}/artifacts/plan.json", plan)
    print(f"✓ Generated plan.json (plan_id: {plan['plan_id'][:8]}..., {len(plan['steps'])} steps)")
    
    trace = generate_trace(context["context_id"], plan["plan_id"])
    write_json(f"{pack_dir}/artifacts/trace.json", trace)
    print(f"✓ Generated trace.json (trace_id: {trace['trace_id'][:8]}..., {len(trace['events'])} events)")
    
    # Generate manifest
    manifest = generate_manifest()
    write_json(f"{pack_dir}/manifest.json", manifest)
    print(f"✓ Generated manifest.json (scenario_id: {manifest['scenario_id']})")
    
    # Generate timeline (NDJSON)
    os.makedirs(f"{pack_dir}/timeline", exist_ok=True)
    with open(f"{pack_dir}/timeline/events.ndjson", "w") as f:
        for event in trace["events"]:
            f.write(json.dumps(event, sort_keys=True) + "\n")
    print(f"✓ Generated timeline/events.ndjson")
    
    # Generate integrity hashes
    sums_path = generate_sha256sums(pack_dir)
    print(f"✓ Generated integrity/sha256sums.txt")
    
    # Calculate pack_root_hash
    with open(sums_path, "rb") as f:
        pack_root_hash = hashlib.sha256(f.read()).hexdigest()
    print(f"\n✅ Pack generated successfully!")
    print(f"   pack_root_hash: {pack_root_hash}")

if __name__ == "__main__":
    main()
